Created guide to connect services using netpol objects (#463)
UKCloud/documentation: articles/openshift/oshift-netpol.md

---
title: How to use NetworkPolicy objects to connect services between projects | UKCloud Ltd
description: Shows how to create a basic NetworkPolicy object to connect services in two projects
services: openshift
author: Kieran O'Neill

toc_rootlink: How To
toc_sub1: 
toc_sub2:
toc_sub3:
toc_sub4:

> toc_title: Using network policy

Reviewer: Change to:
Use NetworkPolicy objects

> toc_fullpath: How To/oshift-netpol.md

Reviewer: Change filename to: oshift-how-use-netpol.md

> toc_mdlink: oshift-netpol.md

Reviewer: Change filename to: oshift-how-use-netpol.md

> ---

Reviewer: Don't forget to change the actual file name!

# How to use NetworkPolicy objects to connect services between projects

## Overview

> In v3.11 clusters, deployed from mid-February 2019, the default SDN plugin has been changed from ovs-multitenant to ovs-networkpolicy. This gives you the ability to create NetworkPolicy objects which allow granular control over the flow of communication between pods/services/projects inside your cluster. 

Reviewer: Do v3.11 clusters deployed before mid-Feb still use ovs-multitenant? If so you need to remove the first comma:

In v3.11 clusters deployed from mid-February 2019, the default SDN plugin...

> In v3.11 clusters, deployed from mid-February 2019, the default SDN plugin has been changed from ovs-multitenant to ovs-networkpolicy. This gives you the ability to create NetworkPolicy objects which allow granular control over the flow of communication between pods/services/projects inside your cluster. 

Reviewer: I'd get rid of the slashes and change to:

...communication between the pods, services and projects inside your cluster.

> By default we put in place NetworkPolicy objects that allow all pods to communicate with each other inside the same project and allow communication from all pods/services to the default and openshift-monitoring projects. This mirrors the default pod security provided with the ovs-multitenant plugin. One of these objects will exist in all projects, including projects created after the cluster is deployed and have the name 'allow-from-same-and-privileged-namespaces'

Reviewer: Change pods/services to pods and services

> By default we put in place NetworkPolicy objects that allow all pods to communicate with each other inside the same project and allow communication from all pods/services to the default and openshift-monitoring projects. This mirrors the default pod security provided with the ovs-multitenant plugin. One of these objects will exist in all projects, including projects created after the cluster is deployed and have the name 'allow-from-same-and-privileged-namespaces'

Reviewer: All projects, including those created after the cluster is deployed, will contain one of these objects, named `allow-from-same-and-privileged-namespaces`.

> By default we put in place NetworkPolicy objects that allow all pods to communicate with each other inside the same project and allow communication from all pods/services to the default and openshift-monitoring projects. This mirrors the default pod security provided with the ovs-multitenant plugin. One of these objects will exist in all projects, including projects created after the cluster is deployed and have the name 'allow-from-same-and-privileged-namespaces'

Reviewer: I found this paragraph a bit confusing. Does the following work?

By default, we create a NetworkPolicy object, named `allow-from-same-and-privileged-namespaces`, in every project, including those created after the cluster is deployed. This object allows all pods to communicate with each other inside the same project and allow communication from all pods and services to the `default` and `openshift-monitoring` projects. This mirrors the default pod security provided with the ovs-multitenant plugin.

Reviewer: Fixed typo in my suggestion

## Interacting with NetworkPolicy objects

> NetworkPolicy objects can interacted with on the command line using the full name networkpolicy or short name netpol, for example:

Reviewer 2: ```suggestion
NetworkPolicy objects can be interacted with on the command line using the full name networkpolicy or short name netpol, for example:
```

> NetworkPolicy objects can be interacted with on the command line using the full name networkpolicy or short name netpol, for example:

Reviewer: You can interact with NetworkPolicy object from the command line using the full name `networkpolicy` or short name `netpol`, for example:

Reviewer: Fixed typo from me

``` oc get netpol ``` - List NetworkPolicy objects in the current project.

> ``` oc describe netpol <name> ``` - Provides more detailed information on the NetworkPolicy object.

Reviewer: on the named NetworkPolicy object.

> ``` oc get netpol <name> -o yaml ``` - Provides yaml output of the NetworkPolicy object. Useful to reference existing rules when creating new objects.

Reviewer: out of the named NetworkPolicy object.

## Example of connecting services from two different projects

> NetworkPolicy objects are scoped at the project level. This means if you create a NetworkPolicy object using a template it will be created inside the project you're in and the rules will apply to pods within that project, with the exception of remote projects you may be referencing inside the rule. The object without any rules will look like the following:

Reviewer: Without any additional rules, the object will look like the following:

Author: I don't think additional rules works here as this is showing the object structure before any rules are added to it

Reviewer: Fair enough

```
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: <desired_name_of_object>
  namespace: <desired_project> (if not provided this will default to the project you are in when you run oc create)
spec:
```

> Inside the spec key we'll pass in the rules that will determine what our NetworkPolicy object does. The first thing you need to consider is the pods you want to target, this is done using the podSelector key. 

Reviewer: Inside the spec key, you can pass in rules that determine what the NetworkPolicy object does.. The first thing you need to consider is the pods you want to target. You can do this using the `podSelector` key.

> Inside the spec field we'll pass in the rules that will determine what our NetworkPolicy object does. The first thing you need to consider is the pods you want to target, this is done using the podSelector field. 

Reviewer 3: ```suggestion
Inside the spec key we'll pass in the rules that will determine what our NetworkPolicy object does. The first thing you need to consider is the pods you want to target, this is done using the podSelector key. 
```

> You can use this to match all pods in a project or alternatively to match all pods with a certain label. If you leave this key blank it will match all pods. If you want to match pods with the label 'role=webserver' 

Reviewer: You can use this to match all pods in a project or alternatively to match all pods with a certain label. To match all pods, leave the key blank. In the example below, we match pods with the label `role=webserver`:

> You can use this to match all pods in a project or alternatively to match all pods with a certain label. If you leave this field blank it will match all pods. If you want to match pods with the label 'role=webserver' 

Reviewer 3: ```suggestion
You can use this to match all pods in a project or alternatively to match all pods with a certain label. If you leave this key blank it will match all pods. If you want to match pods with the label 'role=webserver' 
```

it would look like:

```
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: webserver-egress 
  namespace: webserver
spec:
  podSelector:
    matchLabels:
      role: webserver
```

> At this point we've matched the pods inside our webserver project that have the label role=webserver. Let's say for example we want to allow these pods to connect to a service inside a project we've named 'database' we could

Reviewer: To enable the matched pods to connect to a service inside the pod you can use a namespace selector, which evaluates projects based on a label. You first need to ensure that the appropriate label is applied. The following example applies the namespace label `database=true`:

Reviewer: Fixed my typo

achieve this using a namespace selector, which evaluates projects based on a label. In this case we are going to be looking for the namespace label database=true. Firstly we need to ensure this label is applied using the 
following command:

> [!NOTE]
> The cluster administrator role is required to add labels to namespaces.

``` oc label namespace database database=true ```

> Our netpol object will now look like:

Reviewer: Our example netpol object now looks like the following:

```
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: webserver-egress
  namespace: webserver
spec:
  podSelector:
    matchLabels:
      role: webserver
  egress:
  - to:
    - namespaceSelector:
        matchLabels:
          database: "true"
```

> We now have a template for a NetworkPolicy object that will match any pods labeled role=webserver inside the webserver project and allow them to communicate out to any projects labeled database=true. This is only half of the work,

Reviewer: You now have a template for a NetworkPolicy object that will match...

> We now have a template for a NetworkPolicy object that will match any pods labeled role=webserver inside the webserver project and allow them to communicate out to any projects labeled database=true. This is only half of the work,

Reviewer: We now have a template for a NetworkPolicy object that matches any pods labeled `role=webserver` inside the webserver project and allows them to communicate out to any projects labeled `database=true`. This is only half the work; we still need the ingress allowance inside the database project. Our example template for that object looks like the following:

if we created this object now our desired connectivity would not work as we still need the ingress allowance inside the database project. The template for that object will look like:

```
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: database-ingress
  namespace: database
spec:
  podSelector:
    matchLabels:
      role: database
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          webserver: "true"
```

> This object will be created in the database project, it will match any pods in the project with the label "role=database" and allow ingress traffic from any project labeled "webserver=true". In order to achieve our desired outcome we will need to label our webserver project:

Reviewer: This object is created in the database project and matches any pods in the project with the label `role=database` and allows ingress traffic from any project labeled `webserver=true`. To achieve our desired outcome, we need to label our webserver project as follows:

> This object will be created in the database project, it will match any pods in the project with the label "role=database" and allow ingress traffic from any project labeled "webserver=true" in order to achieve our desired outcome we will need to label our webserver project:

Reviewer 3: ```suggestion
This object will be created in the database project, it will match any pods in the project with the label "role=database" and allow ingress traffic from any project labeled "webserver=true". In order to achieve our desired outcome we will need to label our webserver project:
```

> [!NOTE]
> The cluster administrator role is required to add labels to namespaces.

``` oc label namespace webserver webserver=true ```

> We are now ready to create our objects, in my case the objects have been saved in files called netpol.yaml and netpol1.yaml, substitute the relevant filenames for your environment into the below commands:

Reviewer: Now we can create our ojbects. In our example, we're saving the objects in files called `netpol.yaml` and `netpol1.yaml` using the following commands:

Author: Slightly amended this suggestion to:
In our example, we've saved our templates in files called netpol.yaml and netpol1.yaml. Use the following commands to create the objects, substituting in the relevant filename:

Reviewer: Looks good

``` oc create -f netpol.yaml ```

``` oc create -f netpol1.yaml ```

> Now we have created the objects any pods inside the webserver project matching the label role=webserver will be able to communicate out to the service inside the database project exposed by pods matching the label role=database.

Reviewer: Now we've created the objects, any pods inside the webserver project matching the label `role=webserver` can communicate out to the service inside the database project exposed by pods matching the label `role=database`.

> [!NOTE]

> > You can test connectivity by curling the desired target service from the terminal of a pod which has been allowed access to the service. You can find the terminal in the UI or use ```oc rsh``` to enter the pod.

Reviewer: ...terminal of a pod that has been allowed access...

> > You can test connectivity by curling the desired target service from the terminal of a pod which has been allowed access to the service. You can find the terminal in the UI or use 'oc rsh' to enter the pod.

Reviewer 3: ```suggestion
> You can test connectivity by curling the desired target service from the terminal of a pod which has been allowed access to the service. You can find the terminal in the UI or use `oc rsh` to enter the pod.
```

> ## Further reading

Reviewer: Further information

> In order to get a full understanding of the capabilities and options of a NetworkPolicy object we recommend reading: <https://kubernetes.io/docs/concepts/services-networking/network-policies>

Reviewer: To get a full understanding...

> Documemtation on efficient netpol rules: <https://docs.openshift.com/container-platform/3.11/admin_guide/managing_networking.html#admin-guide-networking-using-networkpolicy-efficiently>

Reviewer 3: ```suggestion
Documentation on efficient netpol rules: <https://docs.openshift.com/container-platform/3.11/admin_guide/managing_networking.html#admin-guide-networking-using-networkpolicy-efficiently>
```
